sandbox/libs/authn/src/main/java/org/opensearch/authn/PermissionHandler.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.authn;
+
+import java.security.Principal;
+import java.util.ArrayList;
+
+/**
+ * This interface represents the abstract notion of a permission handler. A permission handler needs to be able to service the
+ * assignment and
+ *
+ * @opensearch.experimental
+ */
+public interface PermissionHandler {
+
+    // Currently have users --> roles --> permissions want user --> permissions but permissions should be stored elsewhere not directly with
+    // the user objects
+    // Do not need the higher level construct
+    // Want to be able to resolve users to a list of permission
+    // Multi-tenancy is like a private index -- not going to be considered for now.
+    // Each grant having an ID and have a table then the endpoint has the username and a permission -- this grant could end up as a document
+    // in an index
+
+    /**
+     * This function needs to be able to resolve a java Principal to an associated java Subject and return the associated Subject or throw an error if none corresponds
+     */
+    public Subject resolvePrincipal(Principal principal);
```

Comment:
How do you think this would be used? I don't think principals can be turned into subjects without being about to login to the subject, what do you think of this notion?

Reply:
I am still not convinced on needing this. I put it here just in case though, and was imagining we could basically keep a short-lived subject store and then use that to reverse look up the principals associated with each subject. Then you could use the implies(Subject) method to find which Subject a Principal corresponded with. This would let you expend a bit of time on the front to actually grant the permissions all to a Subject using a single Principal. So all my permissions would be under scrawfor99 instead of some being under that and some being under steecraw for instance. We would need to store a pointer to each Subject however and make OpenSearch manage these. They could of course be any given Principal but you would just want to be able to track when a new Principal was used to refer to the same Subject. 

The other option which I prefer is this: Toss the entire resolution step and never resolve a Principal to a Subject. Instead you could just track the permissions granted to principals as Strings or another Object. Then when a Subject tries to take an action you would go through its associated Principals and look for permissions that match what it wants to do to one of its Principals. This is pretty similar to the current Role system in Security and may work nicely.

Reply:
> Toss the entire resolution step and never resolve a Principal to a Subject. 

This option has my vote, 👍